instrumentation/apache-camel-2.20/src/main/java/io/opentelemetry/javaagent/instrumentation/apachecamel/decorators/HttpSpanDecorator.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.javaagent.instrumentation.apachecamel.decorators;
+
+import io.grpc.Context;
+import io.opentelemetry.instrumentation.api.tracer.BaseTracer;
+import io.opentelemetry.javaagent.instrumentation.apachecamel.CamelDirection;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.attributes.SemanticAttributes;
+import java.net.MalformedURLException;
+import java.net.URL;
+import org.apache.camel.Endpoint;
+import org.apache.camel.Exchange;
+
+class HttpSpanDecorator extends BaseSpanDecorator {
+
+  private static final String POST_METHOD = "POST";
+  private static final String GET_METHOD = "GET";
+
+  public static String getHttpMethod(Exchange exchange, Endpoint endpoint) {
+    // 1. Use method provided in header.
+    Object method = exchange.getIn().getHeader(Exchange.HTTP_METHOD);
+    if (method instanceof String) {
+      return (String) method;
+    }
+
+    // 2. GET if query string is provided in header.
+    if (exchange.getIn().getHeader(Exchange.HTTP_QUERY) != null) {
+      return GET_METHOD;
+    }
+
+    // 3. GET if endpoint is configured with a query string.
+    if (endpoint.getEndpointUri().indexOf('?') != -1) {
+      return GET_METHOD;
+    }
+
+    // 4. POST if there is data to send (body is not null).
+    if (exchange.getIn().getBody() != null) {
+      return POST_METHOD;
+    }
+
+    // 5. GET otherwise.
+    return GET_METHOD;
+  }
+
+  @Override
+  public String getOperationName(Exchange exchange, Endpoint endpoint) {
+    // Based on HTTP component documentation:
+    return getHttpMethod(exchange, endpoint);
+  }
+
+  @Override
+  public void pre(Span span, Exchange exchange, Endpoint endpoint, CamelDirection camelDirection) {
+    super.pre(span, exchange, endpoint, camelDirection);
+
+    String httpUrl = getHttpURL(exchange, endpoint);
+    if (httpUrl != null) {
+      span.setAttribute(SemanticAttributes.HTTP_URL, httpUrl);
+    }
+
+    span.setAttribute(SemanticAttributes.HTTP_METHOD, getHttpMethod(exchange, endpoint));
+
+    Span serverSpan = BaseTracer.CONTEXT_SERVER_SPAN_KEY.get(Context.current());
+    if (shouldUpdateServerSpanName(serverSpan, camelDirection)) {
+      updateServerSpanName(serverSpan, httpUrl);
+    }
+  }
+
+  private boolean shouldUpdateServerSpanName(Span serverSpan, CamelDirection camelDirection) {
+
+    return (serverSpan != null && CamelDirection.INBOUND.equals(camelDirection));
+  }
+
+  private void updateServerSpanName(Span serverSpan, String httpUrl) {
+    try {
+      URL url = new URL(httpUrl);
+      serverSpan.updateName(url.getPath());
+    } catch (MalformedURLException e) {
+      // ignored
+    }
```

Comment:
I guess that for server components (`jetty`, `servlet`, ...) it's fine to just use the path, there _shouldn't_ be any path parameters there. But for `rest` component we should use the route - can you make sure that `RestSpanDecorator` uses the correct value here?

Reply:
It's for inbound only - so jetty, servlet, undertow, restlet server. After previous code review i decided to use path (instead of operation name). What do you mean by "route" then ?

Reply:
> What do you mean by "route" then ?

The URI template, like `/resource/{id}/something`

Reply:
But it is - have a look at the `RestSpanDecorator` - it logs "path" of the rest component.